torchdata/stateful_dataloader/stateful_dataloader.py

```diff
@@ -232,27 +235,31 @@ def __iter__(self) -> "_BaseDataLoaderIter":
         # However, in the case of a multiple workers iterator
         # the iterator is only created once in the lifetime of the
         # DataLoader object so that workers can be reused
-        if self.persistent_workers and self.num_workers > 0:
+        if self._initial_iter_for_state_dict:
+            self._initial_iter_for_state_dict = False
+            assert self._iterator is not None
+        elif self.persistent_workers and self.num_workers > 0:
             if self._iterator is None:
                 self._iterator = self._get_iterator()
+                if self._iterator._finished:
+                    self._iterator._reset(self)
             else:
                 self._iterator._reset(self)
         else:
             self._iterator = self._get_iterator()
-        # If we're loading a finished iterator, just request next one immediately
-        if self._iterator._finished:
-            return iter(self)
-        self.iter_calls += 1
+            if self._iterator._finished:
+                self._iterator = self._get_iterator()
         return self._iterator
 
     def state_dict(self) -> Dict[str, Any]:
         if self._iterator is None:
-            return {}
-        else:
-            return self._iterator.state_dict()
+            self._iterator = self._get_iterator()
+            self._initial_iter_for_state_dict = True
+        return self._iterator.state_dict()
 
     def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
         self._iterator = None
+        self._initial_iter_for_state_dict = False
```

Comment:
Actually for your example, I think I'd expect dl_state to be == state, which should be the existing behaviour, and yes setting it False here should ensure that's the case.

the _initial_iter_for_state_dict has a comment in the `__init__` above describing it's usage, but it's basically a micro-optimization to prevent re-creating an iterator if one was created solely for state_dict and persisten_workers = False, we know it was never returned so we can just return it without worrying about reset logic

Reply:
> Actually for your example, I think I'd expect dl_state to be == state, which should be the existing behaviour, and yes setting it False here should ensure that's the case.

I did a quick test: https://pastebin.com/PAxtEdTx and it looks like calling state_dict right after load_state_dict yields an empty dict.

Reply:
sorry by existing behaviour, I meant what's in the code here, not what's in main @gokulavasan